lib/charms/karma_k8s/v0/karma.py

```diff
@@ -0,0 +1,260 @@
+        self.framework.observe(events.relation_departed, self._on_relation_departed)
+        self._stored.set_default(active_relations=set())
+
+    def get_alertmanager_servers(self) -> List[Dict[str, str]]:
+        alertmanager_ips = []
+
+        for relation in self.charm.model.relations[self.name]:
+            if relation.id not in self._stored.active_relations:
+                # relation id is not present in the set of active relations
+                # this probably means that RelationBroken did not exit yet (was recently removed)
+                continue
+
+            # get data from related application
+            data = None
+            for key in relation.data:
+                if key is not self.charm.unit and isinstance(key, ops.charm.model.Unit):
+                    data = relation.data[key]
+                    break
+            if data:
+                config = KarmaAlertmanagerConfig.from_dict(data)
+                if config and config not in alertmanager_ips:
+                    alertmanager_ips.append(config)
+            else:
+                logger.warning("no related units in relation dict")
+
+        return alertmanager_ips  # TODO sorted
+
+    def _on_relation_joined(self, event: RelationJoinedEvent):
+        self._stored.active_relations.add(event.relation.id)
+
+    def _on_relation_changed(self, _):
+        self.on.alertmanager_config_changed.emit()
+
+    def _on_relation_departed(self, event: RelationDepartedEvent):
+        self._stored.active_relations -= {event.relation.id}
+        self.on.alertmanager_config_changed.emit()
+
+    @property
+    def config_valid(self) -> bool:
+        # karma will fail starting without alertmanager server(s), which would cause pebble to error out.
+
+        # check that there is at least one alertmanager server configured
+        servers = self.get_alertmanager_servers()
+        logger.info("config_valid: servers = %s", servers)
+        return len(servers) > 0
+
+
+class KarmaConsumer(ConsumerBase):
+    """A "consumer" handler to be used by charms that relate to Karma (the 'requires' side of the 'karma' relation).
+    This library offers the interface needed in order to provide Alertmanager URIs and associated information to the
+    Karma application.
+
+    To have your charm provide URIs to Karma, declare the interface's use in your charm's metadata.yaml file:
+
+    ```yaml
+    requires:
+      karma-dashboard:
+        interface: karma_dashboard
+    ```
+
+    A typical example of importing this library might be
+
+    ```python
+    from charms.alertmanager_karma.v0.karma import KarmaConsumer
+    ```
+
+    In your charm's `__init__` method:
+
+    ```python
+    self.karma_lib = KarmaConsumer(
+        self,
+        "karma-dashboard",
+        consumes={"karma": ">=0.0.1"},
+    )
+    ```
+
+    The consumer charm is expected to set config via the consumer library, for example in config-changed:
```

Comment:
Perhaps direct the charm writer to where the structure of this "config" is specified.